src/sparse_int_set.jl

```diff
@@ -0,0 +1,292 @@
+
+        packed_id                         = s.reverse[from_page][from_offset]
+        s.packed[packed_id]               = packed_endid
+        s.reverse[to_page][to_offset]     = s.reverse[from_page][from_offset]
+        s.reverse[from_page][from_offset] = 0
+        s.counters[from_page] -= 1
+        pop!(s.packed)
+    end
+    return id
+end
+
+@propagate_inbounds function pop!(s::SparseIntSet, id::Integer)
+    id = dirty_pop!(s, id)
+    cleanup!(s)
+    return id
+end
+
+function pop!(s::SparseIntSet, id::Integer, default)
+    id < 0 && throw(ArgumentError("Int to pop needs to be positive."))
+    in(id, s) ? (@inbounds pop!(s, id)) : default
+end
+popfirst!(s::SparseIntSet) = pop!(s, first(s))
+
+iterate(set::SparseIntSet, args...) = iterate(set.packed, args...) 
+
+last(s::SparseIntSet) = isempty(s) ? throw(ArgumentError("Empty set has no last element.")) : last(s.packed)
+
+union(s::SparseIntSet, ns) = union!(copy(s), ns)
+union!(s::SparseIntSet, ns) = (for n in ns; push!(s, n); end; s)
+
+intersect(s1::SparseIntSet) = copy(s1)
+intersect(s1::SparseIntSet, ss...) = intersect(s1, intersect(ss...))
+function intersect(s1::SparseIntSet, ns)
+    s = SparseIntSet()
+    for n in ns
+        n in s1 && push!(s, n)
+    end
+    return s
+end
+
+intersect!(s1::SparseIntSet, ss...) = intersect!(s1, intersect(ss...))
+
+#Is there a more performant way to do this?
+intersect!(s1::SparseIntSet, ns) = copy!(s1, intersect(s1, ns))
+
+setdiff(s::SparseIntSet, ns) = setdiff!(copy(s), ns)
+setdiff!(s::SparseIntSet, ns) = (for n in ns; pop!(s, n, nothing); end; s)
+
+function ==(s1::SparseIntSet, s2::SparseIntSet)
+    length(s1) != length(s2) && return false
+    return all(x -> in(x, s1), s2)
+end
+
+issubset(a::SparseIntSet, b::SparseIntSet) = isequal(a, intersect(a, b))
+
+complement(a::SparseIntSet) = complement!(SparseIntSet(), a)
+function complement!(b::SparseIntSet, a::SparseIntSet)
+    empty!(b)
+    for i in eachindex(a.reverse)
+        if !isassigned(a.reverse, i)
+            resize!(b.reverse, i)
+            resize!(b.counters, i)
+            new_ids = (i-1)*INT_PER_PAGE+1:(i)*INT_PER_PAGE
+            append!(b.packed, new_ids)
+            b.reverse[i] = collect(new_ids)
+            b.counters[i] = INT_PER_PAGE
+        else
+            for offset in 1:INT_PER_PAGE
+                if a.reverse[i][offset] == 0
+                    push!(b, INT_PER_PAGE*(i-1) + offset)
+                end
+            end
+        end
+    end
+    return b
+end
+#Can this be optimized?
+complement!(a::SparseIntSet) = copy!(a, complement(a))
+
+<(a::SparseIntSet, b::SparseIntSet) = (a<=b) && !isequal(a,b)
+<=(a::SparseIntSet, b::SparseIntSet) = issubset(a, b)
+
+function findfirst_packed_id(i, s::SparseIntSet)
+    pageid, offset = pageid_offset(s, i)
+    if isassigned(s.reverse, pageid)
+        @inbounds id = s.reverse[pageid][offset]
+        return id
+    end
+    return 0
+end
+
+function cleanup!(s::SparseIntSet)
```

Comment:
Here is the version of all my changes in the comments below.

is it correct and the same as current version?

```julia
function cleanup!(s::SparseIntSet)
    last_page_id = findlast(!iszero, s.counters)
    if last_page_id === nothing
        empty!(s.reverse)
        empty!(s.counters)
        return s
    else
        # shrink anything off the end
        resize!(s.counters, last_page_id)
        resize!(s.reverse, last_page_id)

        # Now we need to undef any page in `reverse` that is empty.
        # so we are allocating new pages, and then will copy it across.
        # since can't assign `undef` directly.
        new_pages = Vector{Vector{Int}}(undef, last_page_id)
        for i in eachindex(s.reverse)
            isused =  s.counters[i] !== 0 && isassigned(s.reverse, i)
            if isused
                new_pages[i] = s.reverse[i]
            end
        end
        s.reverse .= new_pages
        return s
    end
end
```